Feature: Added Primal Fear solvers

mariethepolarfox opened the pull request:
## What
Added Math & Public Speaking Primal Fear solvers

## Changelog New Features
+ Added chat solvers for Primal Fears. - Helium9
    * Added solvers for Math and Public Speaking Primal Fears.

## Technical Details
+ Added `generateRandomString` function. - Helium9
+ Fixed spelling in Dev Repo section. - Helium9

mariethepolarfox committed:
Added Primal Fear solvers

Added Math & Public Speaking Primal Fear solvers

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/dev/DebugConfig.java:
```diff
@@ -21,6 +21,11 @@ public class DebugConfig {
     public boolean commandLogs = false;
 
+    @Expose
+    @ConfigOption(name = "Always in SkyBlock", desc = "Enables SkyHanni's features everywhere, not just SkyBlock.")
```
change the wording. since this is in dev category, you can be more technical. something like "this sets skyhannis skylbock detection to always true"

CalMWolfs replied:
I do not like this option as this check is here to stop crashes and other user facing errors. Being an option in the config allows any user to enable it and then they will complain. If a feature needs to work outside of skyblock we already have a place for that. Also I feel that this is so you can use the quickMaths feature outside of skyblock in something like the pit, which id consider a cheat anyways and would not approve for being in skyhanni

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/event/GreatSpookConfig.java:
```diff
@@ -47,4 +48,14 @@ public class GreatSpookConfig {
+    @ConfigOption(name = "Primal Fear Solvers", desc = "Solvers for the Primal Fears")
+    @Accordion
+    @Expose
+    public PrimalFearSolverConfig primalFearSolvers = new PrimalFearSolverConfig();
```
solver or solvers? inconcistent spelling

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/event/GreatSpookConfig.java:
```diff
@@ -47,4 +48,14 @@ public class GreatSpookConfig {
+    // @FeatureToggle
+    // public boolean primalFearSolver = false;
+
+    @ConfigOption(name = "Primal Fear Solvers", desc = "Solvers for the Primal Fears")
```
desc ends with a dot please

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/event/PrimalFearSolverConfig.java:
```diff
@@ -0,0 +1,20 @@
+    @ConfigOption(name = "Math Primal Fear Solver", desc = "Solver for the Math Primal Fear")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean mathPrimalFear = false;
```
nno reason to repeat the name "primal fear" for the variable name in the class that already contains that name

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/event/PrimalFearSolverConfig.java:
```diff
@@ -0,0 +1,20 @@
+
+public class PrimalFearSolverConfig {
+    @Expose
+    @ConfigOption(name = "Math Primal Fear Solver", desc = "Solver for the Math Primal Fear")
```
dot at the end of desc

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/config/features/event/PrimalFearSolverConfig.java:
```diff
@@ -0,0 +1,20 @@
+    public boolean mathPrimalFear = false;
+
+    @Expose
+    @ConfigOption(name = "Public Speaking Primal Fear Solver", desc = "Solver for the Math Primal Fear")
```
check how this title looks in the config. the name is maybe too long and needs shorting

KeinsWolfi commented:
Let me know if you need elaboration on any feature!

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +62,80 @@ object TheGreatSpook {
 
+    private val mathFearMessagePattern by RepoPattern.pattern(
+        "chat.math",
+        "§d§lQUICK MATHS! §r§7Solve: §r§e(?<math>.*)",
```
please add a regex test to this regex pattern. search for "regex-test" in the project for templates to copy from

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +62,80 @@ object TheGreatSpook {
+    )
+
+    private fun solveMath(math: String): String {
+        val mathFormatted = math.replace("x", "*").replace(" ", "")
```
maybe those two replace checks also belong in the math utils?

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +62,80 @@ object TheGreatSpook {
+            val result = MathUtils.evaluateExpression(mathFormatted)
+            result.toString()
+        } catch (e: Exception) {
+            "§c§lError"
```
instead of saying error, return null and handle it custom

mariethepolarfox replied:
I return Strings, how would I return null?

hannibal002 replied:
You change the return value from `String` to `String?`. Then you can replace `"§c§lError"` with `null`. Then you can deal with the nullable in the caller place. (the other function where you were using the function´).

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +62,80 @@ object TheGreatSpook {
+
+        val message = event.message
+
+        if (isMathSolverEnabled()) {
```
you can inline this check.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +62,80 @@ object TheGreatSpook {
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        if (!isSolverEnabled()) return
```
instead check for if in sb here once, then just the config elements below. less confusing functions

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +62,80 @@ object TheGreatSpook {
+    private fun isMathSolverEnabled(): Boolean = LorenzUtils.inSkyBlock && config.primalFearSolvers.mathPrimalFear
+    private fun isSpeakingSolverEnabled(): Boolean = LorenzUtils.inSkyBlock && config.primalFearSolvers.publicSpeakingPrimalFear
+
+    private fun isAllDisabled(): Boolean = !isTimeLeftEnabled() && !isTimerEnabled() && !isFearStatEnabled() &&
```
very confusing

CalMWolfs replied:
To add onyo this, the existing code before you made this feature was also bad but adding another 3 options was the cherry on top. Maybe look at how many other features have a global enabled value and implemetn that instead

mariethepolarfox replied:
Removed entirely as it isnt needed

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/utils/MathUtils.kt:
```diff
@@ -0,0 +1,76 @@
+        return nums[0]
+    }
+
+    private fun String.isDouble(): Boolean {
```
this belongs in number utils. also there might be a better approach to check this, not sure

mariethepolarfox replied:
You already have a isDouble() check in NumberUtils, I removed my redundant function.

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/utils/StringUtils.kt:
```diff
@@ -334,6 +334,11 @@ object StringUtils {
     fun generateRandomId() = UUID.randomUUID().toString()
 
+    fun generateRandomString(length: Int): String {
+        val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
```
store the char array otuside the function so it does not get created on each function call

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/utils/StringUtils.kt:
```diff
@@ -334,6 +334,11 @@ object StringUtils {
 
+    fun generateRandomString(length: Int): String {
+        val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+        return (1..length).map { chars.random() }.joinToString("")
```
look at kotlin's `repeat()` function

hannibal002 reviewed (commented):
just some first thoughts

CalMWolfs commented on src/main/java/at/hannibal2/skyhanni/utils/MathUtils.kt:
This whole file is replaced by NEUCalculator.calculateOrNull()

hannibal002 replied:
neu independency? i see this file as the beginning of a "what if no neu is detected"

CalMWolfs replied:
That is something to deal with before next halloween, but for now does not really matter and I dont see a point in making this new calculator that has a lot less features. 
The best option would be to take the neu calculator and put it into skyhanni like we do with Skyblock time. But this is out of the scope of this pr

mariethepolarfox replied:
I read you want to move away from NEU dependency, which is why I kept mine. The always in Skyblock toggle was mainly for testing reasons in private worlds, if you do not like it, I can remove it again. I will also look at the seccond suggestion

CalMWolfs replied:
> I read you want to move away from NEU dependency

You are correct in this but in this case it is more appropriate to leave it as we already use the neu calculator elsewhere so when we solve that it will also solve this

mariethepolarfox replied:
also removed the isAllDisabled check as it was redundant

mariethepolarfox committed:
Changed code to comply with requests from hannibal

Removed MathUtils as it is not needed at the moment
Fixed typos/spelling

KeinsWolfi commented:
Lmk if you have any other requests!

Obsidianninja11 commented:
I feel like the descriptions aren't very good. Maybe something like `Sends a clickable message in chat to solve the [Math/Public Speaking] Primal Fear.`?
May be out of scope for this pr, but having a keybind for the solvers would be good (and then it could work for the commitment phobia)

mariethepolarfox committed:
Changed wording

Changed wording in the Solver Config
Changed wording in the Public Speaking solver

mariethepolarfox committed:
Added primitive keybindfor Primal Fear Solvers

NEEDS TESTING, DIDNT HAVE THE TIME!!
Added a keybind for quick Solving

KeinsWolfi commented:
NEEDS TESTING!! Im not gonna be home for a bit, can test this evening though.

KeinsWolfi commented:
Reverting keybind now, when reverted it will be ready for review

mariethepolarfox committed:
Removed primitive quick Keybind and finished up both solvers

Removed the primitive Keybind to quickly solve both fears without opening chat
Finished up both solvers

CalMWolfs committed:
Merge branch 'beta' into primalFearSolvers

mariethepolarfox committed:
Code cleanup: Removed unused function

Code cleanup: Removed unused function

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +60,70 @@ object TheGreatSpook {
+                val math = group("math")
+                val result = NEUCalculator.calculateOrNull(math)?.toInt()
+                DelayedRun.runNextTick {
+                    if (result != null) {
```
use `let` instead of `!= null`

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +60,70 @@ object TheGreatSpook {
+                            },
+                        )
+                    } else {
+                        ChatUtils.chat("An unexpected error occurred while solving the math problem.")
```
use error manager and include the math string

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/KeinsWolfi/primalFearSolvers

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -53,11 +60,70 @@ object TheGreatSpook {
     private fun isTimeLeftEnabled(): Boolean = LorenzUtils.inSkyBlock && config.greatSpookTimeLeft
 
-    private fun isAllDisabled(): Boolean = !isTimeLeftEnabled() && !isTimerEnabled() && !isFearStatEnabled() && !isNotificationEnabled()
+    private fun isMathSolverEnabled(): Boolean = LorenzUtils.inSkyBlock && config.primalFearSolver.math
```
remove the two functions: check if in sb only once in the event and inline the two feature checks

mariethepolarfox replied:
only my 2? or every function

hannibal002 committed:
code cleanup

mariethepolarfox committed:
Code Cleanup:
 + Made Bool checks inline
 + Made separate functions for each Primal Fear solver
 + Inlined getting math val

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -44,20 +52,76 @@ object TheGreatSpook {
 
-    private fun isAllDisabled(): Boolean = !isTimeLeftEnabled() && !isTimerEnabled() && !isFearStatEnabled() && !isNotificationEnabled()
+        if (config.primalFearSolver.math) {
+            val math = mathFearMessagePattern.matchMatcher(event.message) { group("math") } ?: return
```
since the logic is moved into mathSolver, the explicit variable math can be removed, reverted to the state above, something like: mathsolver(group("math")))

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -44,20 +52,76 @@ object TheGreatSpook {
+        "§4\\[FEAR] Public Speaking Demon§r§f: (Speak|Say something interesting) (?<name>.*)!",
+    )
+
+    private fun mathSolver(math: String?) {
```
maybe rename math to query, term, question, task, or something. math is too abstract imo

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -44,20 +52,76 @@ object TheGreatSpook {
+            return
+        }
+        ChatUtils.clickToActionOrDisable(
+            "The result is: $result",
```
maybe rename result to answer. not sure though

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/event/spook/TheGreatSpook.kt:
```diff
@@ -44,20 +52,76 @@ object TheGreatSpook {
-    private fun isTimerEnabled(): Boolean = LorenzUtils.inSkyBlock && config.primalFearTimer
+    private fun publicSpeakingSolver() {
+        ChatUtils.clickToActionOrDisable(
+            "Click to send a random string to complete the Primal Fear",
```
remove the phrase "to send a random string" from this line. otherwise the line gets too long also the whole chat message contains the information twice atm

mariethepolarfox committed:
Code Cleanup

Code Cleanup

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/KeinsWolfi/primalFearSolvers

hannibal002 committed:
code cleanup
